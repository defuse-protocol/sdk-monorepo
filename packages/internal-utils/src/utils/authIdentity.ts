import { keccak_256 } from "@noble/hashes/sha3";
import { base58, hex } from "@scure/base";
import { Keypair } from "@stellar/stellar-sdk";
import type {
	AuthHandle,
	AuthIdentifier,
	AuthMethod,
} from "../types/authHandle";
import type { IntentsUserId } from "../types/intentsUserId";
import { assert } from "./assert";
import { parsePublicKey } from "./webAuthn";

/**
 * Converts a blockchain address to a standardized Defuse user ID.
 *
 * The conversion follows these rules:
 * 1. NEAR addresses: Used as-is (lowercased)
 *   - Explicit: "bob.near"
 *   - Implicit: "17208628f84f5d6ad33f0da3bbbeb27ffcb398eac501a31bd6ad2011e36133a1"
 *
 * 2. EVM addresses: Used as-is (lowercased)
 *   - Format: "0xc0ffee254729296a45a3885639ac7e10f9d54979"
 *
 * 3. Solana addresses: Converted from base58 to hex
 *   - Input: base58 public key
 *   - Output: hex-encoded string
 *
 * 4. WebAuthn credentials: Converted based on curve type
 *   - P-256: Keccak256(prefix + pubkey) -> last 20 bytes -> hex with 0x prefix
 *   - Ed25519: Raw public key -> hex encoding
 *
 * @param authIdentifier - The user's identifier (blockchain address or WebAuthn public key)
 * @param authMethod - The type of credential ("evm", "near", "solana", "webauthn")
 * @returns A standardized Defuse user ID
 */
export function authHandleToIntentsUserId(
	authIdentifier: AuthIdentifier,
	authMethod: AuthMethod,
): IntentsUserId;
export function authHandleToIntentsUserId(
	authHandle: AuthHandle,
): IntentsUserId;
export function authHandleToIntentsUserId(
	authIdentifier: AuthIdentifier | AuthHandle,
	authMethod?: AuthMethod,
): IntentsUserId {
	let authHandle: AuthHandle;
	if (typeof authIdentifier === "object") {
		authHandle = authIdentifier;
	} else if (authMethod != null) {
		authHandle = {
			identifier: authIdentifier,
			method: authMethod,
		};
	} else {
		// This should never happen, because of argument types
		throw new Error("Invalid arguments");
	}

	const method = authHandle.method;
	switch (method) {
		case "evm":
		case "near":
			return authHandle.identifier.toLowerCase() as IntentsUserId;

		case "solana":
			return hex.encode(base58.decode(authHandle.identifier)) as IntentsUserId;

		case "webauthn": {
			return webAuthnIdentifierToIntentsUserId(
				authHandle.identifier,
			) as IntentsUserId;
		}

		case "ton": {
			assert(authHandle.identifier.length === 64);
			hex.decode(authHandle.identifier);
			return authHandle.identifier.toLowerCase() as IntentsUserId;
		}

		case "stellar": {
			const keypair = Keypair.fromPublicKey(authHandle.identifier);
			return hex.encode(keypair.rawPublicKey()) as IntentsUserId;
		}

		default:
			method satisfies never;
			throw new Error("Unsupported auth method");
	}
}

function webAuthnIdentifierToIntentsUserId(credential: string): string {
	const { curveType, publicKey } = parsePublicKey(credential);

	switch (curveType) {
		case "p256": {
			const p256 = new TextEncoder().encode("p256");
			const addressBytes = keccak_256(
				new Uint8Array([...p256, ...publicKey]),
			).slice(-20);

			// biome-ignore lint/style/useTemplate: it's fine
			return "0x" + hex.encode(addressBytes);
		}

		case "ed25519": {
			return hex.encode(publicKey);
		}

		default:
			curveType satisfies never;
			throw new Error("Unsupported curve type");
	}
}
