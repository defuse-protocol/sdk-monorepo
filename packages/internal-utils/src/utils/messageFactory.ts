import type {
	Intent,
	Nep413DefuseMessageFor_DefuseIntents,
} from "@defuse-protocol/contract-types";
import { sha256 } from "@noble/hashes/sha256";
import { base64 } from "@scure/base";
import { getAddress } from "viem";
import { config } from "../config";
import { createWithdrawMemo } from "../poaBridge/createWithdrawMemo";
import type { SupportedChainName } from "../types/base";
import type { IntentsUserId } from "../types/intentsUserId";
import type { WalletMessage } from "../types/walletMessage";
import { assert } from "./assert";
import { buildHotOmniWithdrawIntent } from "./hotOmniUtils";

/**
 * @param tokenDeltas
 * @param signerId
 * @param deadlineTimestamp Unix timestamp in milliseconds
 * @param referral
 * @param memo
 * @param appFee
 * @param appFeeRecipient
 */
export function makeInnerSwapMessage({
	tokenDeltas,
	signerId,
	deadlineTimestamp,
	referral,
	memo,
	appFee,
	appFeeRecipient,
}: {
	tokenDeltas: [string, bigint][];
	signerId: IntentsUserId;
	deadlineTimestamp: number;
	referral?: string;
	memo?: string;
	appFee: [string, bigint][];
	appFeeRecipient: string;
}): Nep413DefuseMessageFor_DefuseIntents {
	const tokenDiff: Record<string, string> = {};
	const tokenDiffNum: Record<string, bigint> = {};

	const intents: Intent[] = [
		{
			intent: "token_diff",
			diff: tokenDiff,
			referral,
			memo,
		},
	];

	if (appFee.length) {
		intents.push({
			intent: "transfer",
			tokens: Object.fromEntries(
				appFee.map(([token, amount]) => [token, amount.toString()]),
			),
			receiver_id: appFeeRecipient,
			memo: "APP_FEE",
		});
	}

	for (const [token, amount] of tokenDeltas) {
		tokenDiffNum[token] ??= 0n;
		tokenDiffNum[token] += amount;
		// biome-ignore lint/style/noNonNullAssertion: it is checked above
		tokenDiff[token] = tokenDiffNum[token]!.toString();
	}

	if (Object.keys(tokenDiff).length === 0) {
		return {
			deadline: new Date(deadlineTimestamp).toISOString(),
			intents: [],
			signer_id: signerId,
		};
	}

	return {
		deadline: new Date(deadlineTimestamp).toISOString(),
		intents,
		signer_id: signerId,
	};
}

/**
 * Explanation of `tokenDelta` and `storageTokenDelta`.
 *
 * Say we withdraw token $Y, but we have token $X. We need to swap $X to $Y first.
 * `tokenDelta` represents the swap from $X to $Y.
 *
 * Say we withdraw token $Y to Near, and we need to pay for storage.
 * Storage is paid in token $N.
 * `storageTokenDelta` represents the swap from $Y to $N.
 *
 * Important: We must execute these swaps separately rather than combining them.
 * This is because protocol fees are applied to each outgoing token in a swap.
 *
 * Example of why combining swaps fails when fee is 0.3%:
 *
 * User withdraws 1.0 USDC:eth to Near blockchain and pays for storage.
 * Quote #1: 1.0 USDC:eth -> 0.994009 USDC:near
 * Quote #2: 0.003866 USDC:near -> 0.00125 NEAR
 *
 * User combined diff: [-1000000 USDC:eth, +994009-3866=+990143 USDC:near, +125000 NEAR]
 *   - Shared pool state: [+997000 USDC:eth, -990143 USDC:near, -125000 NEAR]
 *   - Note: 1000000*0.003=3000 USDC:eth taken as protocol fee
 *
 * Solver #1 diff: [+997000 USDC:eth, -997000 USDC:near, 0 NEAR]
 *   - Shared pool state: [+997000-997000=0 USDC:eth, -990143+994009=+3866 USDC:near, -125000 NEAR]
 *
 * Solver #2 diff: [0 USDC:eth, +3854 USDC:near, -125377 NEAR]
 *  - Shared pool state: [0 USDC:eth, +3866-3854=12 USDC:near, -125000+125000=0 NEAR]
 *
 * Result: 12 USDC:near remains in the pool - violating the invariant that at the end the pool should be empty.
 *
 * By executing swaps separately, we properly account for protocol fees at each step.
 *
 * @param tokenDeltas - the swap from X to Y, where Y to be withdrawn
 * @param storageTokenDeltas - the swap from Y to N, where Y to be withdrawn, N to be paid for storage
 * @param withdrawParams
 * @param signerId
 * @param deadlineTimestamp - unix timestamp in seconds
 * @param referral
 */
export function makeInnerSwapAndWithdrawMessage({
	tokenDeltas,
	storageTokenDeltas,
	withdrawParams,
	signerId,
	deadlineTimestamp,
	referral,
}: {
	tokenDeltas: [string, bigint][];
	storageTokenDeltas: [string, bigint][];
	withdrawParams: WithdrawParams;
	signerId: IntentsUserId;
	deadlineTimestamp: number;
	referral?: string;
}): Nep413DefuseMessageFor_DefuseIntents {
	const intents: NonNullable<Nep413DefuseMessageFor_DefuseIntents["intents"]> =
		[];

	if (tokenDeltas.length) {
		const { intents: swapIntents } = makeInnerSwapMessage({
			tokenDeltas,
			signerId,
			deadlineTimestamp,
			referral,
			appFee: [],
			appFeeRecipient: "",
		});
		assert(swapIntents, "swapIntents must be defined");
		intents.push(...swapIntents);
	}

	if (storageTokenDeltas.length) {
		const { intents: storageIntents } = makeInnerSwapMessage({
			tokenDeltas: storageTokenDeltas,
			signerId,
			deadlineTimestamp,
			referral,
			appFee: [],
			appFeeRecipient: "",
		});
		assert(storageIntents, "storageIntents must be defined");
		intents.push(...storageIntents);
	}

	intents.push(makeInnerWithdrawMessage(withdrawParams));

	return {
		deadline: new Date(deadlineTimestamp).toISOString(),
		intents: intents,
		signer_id: signerId,
	};
}

export type WithdrawParams =
	| {
			type: "to_near";
			amount: bigint;
			tokenAccountId: string;
			receiverId: string;
			storageDeposit: bigint;
	  }
	| {
			type: "via_poa_bridge";
			amount: bigint;
			tokenAccountId: string;
			destinationAddress: string;
			destinationMemo: string | null;
	  }
	| {
			type: "to_aurora_engine";
			amount: bigint;
			tokenAccountId: string;
			auroraEngineContractId: string;
			destinationAddress: string;
	  }
	| {
			type: "hot_omni";
			chainName: SupportedChainName;
			amount: bigint;
			defuseAssetId: string;
			destinationAddress: string; // todo: consider renaming `receiverId` and `destinationAddress` to `recipient`?
	  };

function makeInnerWithdrawMessage(params: WithdrawParams): Intent {
	const paramsType = params.type;
	switch (paramsType) {
		case "to_near":
			if (params.tokenAccountId === "wrap.near") {
				return {
					intent: "native_withdraw",
					receiver_id: params.receiverId,
					amount: params.amount.toString(),
				};
			}
			return {
				intent: "ft_withdraw",
				token: params.tokenAccountId,
				receiver_id: params.receiverId,
				amount: params.amount.toString(),
				storage_deposit:
					params.storageDeposit > 0n ? params.storageDeposit.toString() : null,
			};

		case "via_poa_bridge": {
			return {
				intent: "ft_withdraw",
				token: params.tokenAccountId,
				receiver_id: params.tokenAccountId,
				amount: params.amount.toString(),
				memo: createWithdrawMemo({
					receiverAddress: params.destinationAddress,
					xrpMemo: params.destinationMemo,
				}),
			};
		}

		case "to_aurora_engine":
			return {
				intent: "ft_withdraw",
				token: params.tokenAccountId,
				receiver_id: params.auroraEngineContractId,
				amount: params.amount.toString(),
				msg: makeAuroraEngineDepositMsg(params.destinationAddress),
			};

		case "hot_omni": {
			return buildHotOmniWithdrawIntent({
				chainName: params.chainName,
				defuseAssetId: params.defuseAssetId,
				amount: params.amount,
				receiver: params.destinationAddress,
			});
		}

		default:
			paramsType satisfies never;
			throw new Error(`Unknown withdraw type: ${paramsType}`);
	}
}

export function makeSwapMessage({
	innerMessage,
	nonce = randomDefuseNonce(),
}: {
	innerMessage: Nep413DefuseMessageFor_DefuseIntents;
	nonce?: Uint8Array;
}): WalletMessage {
	const payload = {
		signer_id: innerMessage.signer_id,
		verifying_contract: config.env.contractID,
		deadline: innerMessage.deadline,
		nonce: base64.encode(nonce),
		intents: innerMessage.intents,
	};
	const payloadSerialized = JSON.stringify(payload);
	const payloadBytes = new TextEncoder().encode(payloadSerialized);

	return {
		NEP413: {
			message: JSON.stringify(innerMessage),
			// This is who will be verifying the message
			recipient: config.env.contractID,
			nonce,
		},
		ERC191: {
			message: JSON.stringify(payload, null, 2),
		},
		SOLANA: {
			message: payloadBytes,
		},
		STELLAR: {
			message: JSON.stringify(payload, null, 2),
		},
		WEBAUTHN: {
			challenge: makeChallenge(payloadBytes),
			payload: payloadSerialized,
			parsedPayload: payload,
		},
		TON_CONNECT: {
			message: {
				type: "text",
				text: JSON.stringify(payload, null, 2),
			},
		},
	};
}

export function makeEmptyMessage({
	signerId,
	deadlineTimestamp,
	nonce = randomDefuseNonce(),
}: {
	signerId: IntentsUserId;
	deadlineTimestamp: number;
	nonce?: Uint8Array;
}): WalletMessage {
	const innerMessage: Nep413DefuseMessageFor_DefuseIntents = {
		deadline: new Date(deadlineTimestamp).toISOString(),
		intents: [],
		signer_id: signerId,
	};

	return makeSwapMessage({
		innerMessage,
		nonce,
	});
}

export function randomDefuseNonce(): Uint8Array {
	return randomBytes(32);
}

function randomBytes(length: number): Uint8Array {
	return crypto.getRandomValues(new Uint8Array(length));
}

/**
 * In order to deposit to AuroraEngine powered chain, we need to have a `msg`
 * with the destination address in special format (lower case + without 0x).
 */
function makeAuroraEngineDepositMsg(recipientAddress: string): string {
	const parsedRecipientAddress = getAddress(recipientAddress);
	return parsedRecipientAddress.slice(2).toLowerCase();
}

/**
 * Converts UTF-8 string to bytes for WebAuthn challenge
 */
export function makeChallenge(payload: Uint8Array): Uint8Array {
	// It's possible to use native crypto, but it's async, and this would break existing flow:
	// await crypto.subtle.digest("SHA-256", messageBytes)
	const hash = sha256(payload);
	return new Uint8Array(hash);
}

export function makeInnerTransferMessage({
	tokenDeltas,
	signerId,
	deadlineTimestamp,
	receiverId,
	memo,
}: {
	tokenDeltas: [string, bigint][];
	signerId: IntentsUserId;
	deadlineTimestamp: number;
	receiverId: string;
	memo?: string;
}): Nep413DefuseMessageFor_DefuseIntents {
	const tokens: Record<string, string> = {};
	const seenTokens = new Set<string>();

	for (const [token, amount] of tokenDeltas) {
		assert(!seenTokens.has(token), `Duplicate token found: ${token}`);
		seenTokens.add(token);
		assert(
			amount > 0n,
			`Transfer amount must be positive, got: ${amount} for token ${token}`,
		);
		tokens[token] = amount.toString();
	}

	return {
		deadline: new Date(deadlineTimestamp).toISOString(),
		intents: [
			{
				intent: "transfer",
				tokens,
				receiver_id: receiverId,
				...(memo ? { memo } : {}),
			},
		],
		signer_id: signerId,
	};
}
