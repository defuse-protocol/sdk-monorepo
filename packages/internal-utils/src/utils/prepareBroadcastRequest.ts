import { base58, base64, hex } from "@scure/base";
import type {
	Params,
	PublishIntentRequest,
} from "../solverRelay/solverRelayHttpClient/types";
import type { AuthMethod } from "../types/authHandle";
import type { WalletSignatureResult } from "../types/walletMessage";
import { assert } from "./assert";
import { makeWebAuthnMultiPayload } from "./multiPayload/webauthn";
import { stellarAddressToBytes } from "./stellarAddressToBytes";

export function prepareSwapSignedData(
	signature: WalletSignatureResult,
	userInfo: { userAddress: string; userChainType: AuthMethod },
): Params<PublishIntentRequest>["signed_data"] {
	const signatureType = signature.type;
	switch (signatureType) {
		case "NEP413": {
			return {
				standard: "nep413",
				payload: {
					message: signature.signedData.message,
					nonce: base64.encode(signature.signedData.nonce),
					recipient: signature.signedData.recipient,
					callbackUrl: signature.signedData.callbackUrl,
				},
				public_key: signature.signatureData.publicKey, // publicKey is already in the correct format
				signature: transformNEP141Signature(signature.signatureData.signature),
			};
		}

		case "ERC191": {
			return {
				standard: "erc191",
				payload: signature.signedData.message,
				signature: transformERC191Signature(signature.signatureData),
			};
		}

		case "SOLANA":
			assert(
				userInfo.userChainType === "solana",
				"User chain and signature chain must match",
			);
			return {
				standard: "raw_ed25519",
				payload: new TextDecoder().decode(signature.signedData.message),
				// Solana address is its public key encoded in base58
				public_key: `ed25519:${userInfo.userAddress}`,
				signature: transformED25519Signature(signature.signatureData),
			};

		case "WEBAUTHN": {
			return makeWebAuthnMultiPayload(userInfo, signature);
		}

		case "TON_CONNECT": {
			return {
				standard: "ton_connect",
				address: signature.signatureData.address,
				domain: signature.signatureData.domain,
				timestamp: signature.signatureData.timestamp,
				payload: signature.signatureData.payload,
				public_key: `ed25519:${base58.encode(hex.decode(userInfo.userAddress))}`,
				signature: `ed25519:${base58.encode(base64.decode(signature.signatureData.signature))}`,
			};
		}

		case "STELLAR_SEP53": {
			assert(
				userInfo.userChainType === "stellar",
				"User chain and signature chain must match",
			);
			return {
				standard: "sep53",
				payload: signature.signedData.message,
				// We should encode the Stellar address to base58
				public_key: `ed25519:${base58.encode(
					stellarAddressToBytes(userInfo.userAddress),
				)}`,
				signature: transformED25519Signature(signature.signatureData),
			};
		}

		case "TRON": {
			return {
				standard: "tip191",
				payload: signature.signedData.message,
				signature: transformERC191Signature(signature.signatureData), // TIP-191 is compatible with ERC191
			};
		}

		default:
			signatureType satisfies never;
			throw new Error("exhaustive check failed");
	}
}

function transformNEP141Signature(signature: string) {
	const encoded = base58.encode(base64.decode(signature));
	return `ed25519:${encoded}`;
}

export function transformERC191Signature(signature: string) {
	const normalizedSignature = normalizeERC191Signature(signature);
	const bytes = hex.decode(
		normalizedSignature.startsWith("0x")
			? normalizedSignature.slice(2)
			: normalizedSignature,
	);
	return `secp256k1:${base58.encode(bytes)}`;
}

export function normalizeERC191Signature(signature: string): string {
	// Get `v` from the last two characters
	let v = Number.parseInt(signature.slice(-2), 16);

	// // Normalize `v` to be either 0 or 1
	v = toRecoveryBit(v);

	// Convert `v` back to hex
	const vHex = v.toString(16).padStart(2, "0");

	// Reconstruct the full signature with the adjusted `v`
	return signature.slice(0, -2) + vHex;
}

// Copy from viem/utils/signature/recoverPublicKey.ts
function toRecoveryBit(yParityOrV: number) {
	if (yParityOrV === 0 || yParityOrV === 1) return yParityOrV;
	if (yParityOrV === 27) return 0;
	if (yParityOrV === 28) return 1;
	throw new Error("Invalid yParityOrV value");
}

function transformED25519Signature(signature: Uint8Array) {
	return `ed25519:${base58.encode(signature)}`;
}
