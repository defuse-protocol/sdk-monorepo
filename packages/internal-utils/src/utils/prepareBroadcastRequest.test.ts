import { describe, expect, it } from "vitest";
import type {
	ERC191SignatureData,
	NEP413SignatureData,
	SolanaSignatureData,
	StellarSignatureData,
	TonConnectSignatureData,
	WalletMessage,
	WebAuthnSignatureData,
} from "../types/walletMessage";
import { authHandleToIntentsUserId } from "./authIdentity";
import { makeInnerSwapMessage, makeSwapMessage } from "./messageFactory";
import { prepareSwapSignedData } from "./prepareBroadcastRequest";

describe("prepareSwapSignedData()", () => {
	const swapMessage = makeSwapMessage({
		innerMessage: makeInnerSwapMessage({
			tokenDeltas: [["foo.near", 100n]],
			signerId: authHandleToIntentsUserId("user.near", "near"),
			deadlineTimestamp: 1704110400000,
			appFee: [],
			appFeeRecipient: "",
		}),
		nonce: new Uint8Array(32),
	});

	const walletMessage: WalletMessage = {
		NEP413: {
			message: `{"foo":"bar"}`,
			recipient: "defuse.near",
			nonce: Buffer.from(
				"esXbbxyJNApGznX1v8kT5ojuat7jqUv84Ib+Q6hWdzI=",
				"base64",
			),
		},
		ERC191: {
			message: JSON.stringify({ foo: "bar" }),
		},
		SOLANA: {
			message: Uint8Array.from(
				Buffer.from(JSON.stringify({ foo: "bar" }), "utf8"),
			),
		},
		STELLAR: {
			message: JSON.stringify({ foo: "bar" }),
		},
		WEBAUTHN: swapMessage.WEBAUTHN,
		TON_CONNECT: swapMessage.TON_CONNECT,
		TRON: {
			message: JSON.stringify({ foo: "bar" }),
		},
	};

	it("should return the correct signed data for a NEP141 signature", () => {
		const signature: NEP413SignatureData = {
			type: "NEP413",
			signatureData: {
				accountId: "user.near",
				publicKey: "ed25519:Gxa24TGbJu4mqdhW3GbvLXmf4bSEyxVicrtpChDWbgga",
				signature: "stH6JnShG+GwWCsfN9iu/m4un6qwYLN9Df+5oQYsL7Q=",
			},
			signedData: walletMessage.NEP413,
		};

		expect(
			prepareSwapSignedData(signature, {
				userAddress: "user.near",
				userChainType: "near",
			}),
		).toMatchSnapshot();
	});

	it("should return the correct signed data for an ERC191 signature", () => {
		const signature: ERC191SignatureData = {
			type: "ERC191",
			signatureData: "0xdeadbeef1c",
			signedData: walletMessage.ERC191,
		};

		expect(
			prepareSwapSignedData(signature, {
				userAddress: "0xabcd",
				userChainType: "evm",
			}),
		).toMatchSnapshot();
	});

	it("should return the correct signed data for a Solana signature", () => {
		const signature: SolanaSignatureData = {
			type: "SOLANA",
			signatureData: Buffer.from("deadbeef1c", "hex"),
			signedData: walletMessage.SOLANA,
		};

		expect(
			prepareSwapSignedData(signature, {
				userAddress: "DRpbCBMxVnDK7maPM5tGv6MvB3v1sRMC86PZ8okm21hy",
				userChainType: "solana",
			}),
		).toMatchSnapshot();
	});

	it("should return the correct signed data for a Stellar SEP-0053 signature", () => {
		const signature: StellarSignatureData = {
			type: "STELLAR_SEP53",
			signatureData: Buffer.from("deadbeef1c", "hex"),
			signedData: {
				message: JSON.stringify({ foo: "bar" }),
			},
		};

		expect(
			prepareSwapSignedData(signature, {
				userAddress: "GDZ7T36V5BBWLZTM7NGZYWJYRIYBFXFVYQT3EOGO7G4JCEO6DBYL7DC2",
				userChainType: "stellar",
			}),
		).toMatchSnapshot();
	});

	it.skip("should return the correct signed data for a WebAuthn signature", async () => {
		const signature: WebAuthnSignatureData = {
			type: "WEBAUTHN",
			signatureData: {
				authenticatorData: Buffer.from("dead", "hex").buffer,
				clientDataJSON: Buffer.from('{"some": "json"}', "utf-8").buffer,
				signature: Buffer.from("beef", "hex").buffer,
				userHandle: Buffer.from("1ee7", "hex").buffer,
			},
			signedData: walletMessage.WEBAUTHN,
		};

		expect(
			prepareSwapSignedData(signature, {
				userAddress: "ed25519:Gxa24TGbJu4mqdhW3GbvLXmf4bSEyxVicrtpChDWbgga",
				userChainType: "webauthn",
			}),
		).toMatchSnapshot();
	});

	it("should return the correct signed data for a TonConnect signature", () => {
		const signature: TonConnectSignatureData = {
			type: "TON_CONNECT",
			signatureData: {
				signature:
					"4o7K0k+5pQUHVTX3fD03JIwcixdxmPE7pGgVmQsUHxdZ+G2OJEeXKDv5cnrgPPbZQDgUrMGWfXhYvRFpdJxuAg==",
				address:
					"0:fa63f5195b0f8682d3f3413e2b40decfae7778b3691748a2d55dae5b243a3054",
				timestamp: 1748949269,
				domain: "ton-connect.github.io",
				payload: walletMessage.TON_CONNECT.message,
			},
			signedData: walletMessage.TON_CONNECT,
		};

		expect(
			prepareSwapSignedData(signature, {
				userAddress:
					"d1e7c122f8a43c7d7433548c4604edd4dffcfe5bb1d036499684980c115500bf",
				userChainType: "ton",
			}),
		).toMatchSnapshot();
	});
});
