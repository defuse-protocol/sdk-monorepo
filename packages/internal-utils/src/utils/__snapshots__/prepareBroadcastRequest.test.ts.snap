// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prepareSwapSignedData() > should return the correct signed data for a NEP141 signature 1`] = `
{
  "payload": {
    "callbackUrl": undefined,
    "message": "{"foo":"bar"}",
    "nonce": "esXbbxyJNApGznX1v8kT5ojuat7jqUv84Ib+Q6hWdzI=",
    "recipient": "defuse.near",
  },
  "public_key": "ed25519:Gxa24TGbJu4mqdhW3GbvLXmf4bSEyxVicrtpChDWbgga",
  "signature": "ed25519:D33GxHszz1ZpkXpnEWYAwkQxeYBU9YCkTMKZ5uwDS1B9",
  "standard": "nep413",
}
`;

exports[`prepareSwapSignedData() > should return the correct signed data for a Solana signature 1`] = `
{
  "payload": "{"foo":"bar"}",
  "public_key": "ed25519:DRpbCBMxVnDK7maPM5tGv6MvB3v1sRMC86PZ8okm21hy",
  "signature": "ed25519:S88bG8B",
  "standard": "raw_ed25519",
}
`;

exports[`prepareSwapSignedData() > should return the correct signed data for a Stellar signature 1`] = `
{
  "payload": "{"foo":"bar"}",
  "public_key": "ed25519:274mMrvWU1KsV8JWMw3ny9cP66fbsZL7vkzZtENBrBnM",
  "signature": "ed25519:S88bG8B",
  "standard": "raw_ed25519",
}
`;

exports[`prepareSwapSignedData() > should return the correct signed data for a TonConnect signature 1`] = `
{
  "address": "0:fa63f5195b0f8682d3f3413e2b40decfae7778b3691748a2d55dae5b243a3054",
  "domain": "ton-connect.github.io",
  "payload": {
    "text": "{
  "signer_id": "user.near",
  "verifying_contract": "intents.near",
  "deadline": "2024-01-01T12:00:00.000Z",
  "nonce": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
  "intents": [
    {
      "intent": "token_diff",
      "diff": {
        "foo.near": "100"
      }
    }
  ]
}",
    "type": "text",
  },
  "public_key": "ed25519:F8PB56zdMYNDL7Mq43DV4cV17uRqQkpn6ZygNdqavXCr",
  "signature": "ed25519:5XidPDRhEdQBvHuMx9xAx4q6hfsBzYtv4XDxpsfDgTmSXgZhEboy7kytdQfwJCKnjLYQwPrdqc7ACR3tr6F1HSGR",
  "standard": "ton_connect",
  "timestamp": 1748949269,
}
`;

exports[`prepareSwapSignedData() > should return the correct signed data for an ERC191 signature 1`] = `
{
  "payload": "{"foo":"bar"}",
  "signature": "secp256k1:S88bG7i",
  "standard": "erc191",
}
`;
